tests/lib/rules/template-tag-spacing.js

```diff
@@ -0,0 +1,259 @@
+ * @author Jonathan Wilsson
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const rule = require("../../../lib/rules/template-tag-spacing"),
+    RuleTester = require("../../../lib/testers/rule-tester");
+
+//------------------------------------------------------------------------------
+// Tests
+//------------------------------------------------------------------------------
+
+const ruleTester = new RuleTester({ parserOptions: { ecmaVersion: 6 } });
+
+ruleTester.run("template-tag-spacing", rule, {
+    valid: [
+        "tag`name`",
+        { code: "tag`name`", options: ["never"] },
+        { code: "tag `name`", options: ["always"] },
+        "tag`hello ${name}`",
+        { code: "tag`hello ${name}`", options: ["never"] },
+        { code: "tag `hello ${name}`", options: ["always"] },
+        "tag/*here's a comment*/`Hello world`",
+        { code: "tag/*here's a comment*/`Hello world`", options: ["never"] },
+        { code: "tag /*here's a comment*/`Hello world`", options: ["always"] },
+        { code: "tag/*here's a comment*/ `Hello world`", options: ["always"] },
+        "new tag`name`",
+        { code: "new tag`name`", options: ["never"] },
+        { code: "new tag `name`", options: ["always"] },
+        "new tag`hello ${name}`",
+        { code: "new tag`hello ${name}`", options: ["never"] },
+        { code: "new tag `hello ${name}`", options: ["always"] },
+        "(tag)`name`",
+        { code: "(tag)`name`", options: ["never"] },
+        { code: "(tag) `name`", options: ["always"] },
+        "(tag)`hello ${name}`",
+        { code: "(tag)`hello ${name}`", options: ["never"] },
+        { code: "(tag) `hello ${name}`", options: ["always"] },
+        "new (tag)`name`",
+        { code: "new (tag)`name`", options: ["never"] },
+        { code: "new (tag) `name`", options: ["always"] },
+        "new (tag)`hello ${name}`",
+        { code: "new (tag)`hello ${name}`", options: ["never"] },
+        { code: "new (tag) `hello ${name}`", options: ["always"] }
+    ],
+    invalid: [
+        {
+            code: "tag `name`",
+            output: "tag`name`",
+            errors: [
+                { message: "Unexpected space between template tag and template literal." }
+            ]
+        },
+        {
+            code: "tag `name`",
+            output: "tag`name`",
+            errors: [
+                { message: "Unexpected space between template tag and template literal." }
+            ],
+            options: ["never"]
+        },
+        {
+            code: "tag`name`",
+            output: "tag `name`",
+            errors: [
+                { message: "Missing space between template tag and template literal." }
+            ],
+            options: ["always"]
+        },
+        {
+            code: "tag /*here's a comment*/`Hello world`",
```

Comment:
I think the best way to fix this would be to get a list of all the comments between the tag and the template literal, then concatenate them to get the resulting text between the two tokens. This would preserve the whitespace in the comments themselves and remove any other whitespace.

However, we should make sure we don't change the behavior of this code:

```js
foo // line comment
`bar`
```

The implementation I mentioned above would fix that to

```js
foo// line comment`bar`
```

To avoid this, I think we should avoid doing a fix if there are any line comments between the tag and the template literal.

So I would do something along the lines of

```js
context.report({
  // ...
  fix(fixer) {
    const comments = sourceCode.getComments(node.quasi).leading;

    if (comments.some(comment => comment.type === "Line")) {
        return null;
    }
    return fixer.replaceTextRange(
        [tagToken.range[1], literalToken.range[0]],
        comments.reduce((text, comment) => text + sourceCode.getText(comment), "")
    );
  }
});
```